### DISCUSSION TITLE
Bevy input lag vs egui for GUIs

### DISCUSSION THREAD

**Discussion by DKolter** [bevyengine/bevy#7599] [Q&A] [answered]
Yesterday I tried to port a tile editor GUI from egui to bevy using the bevy egui plugin as a replacement, because I hoped to get a performance boost. The GPU usage is significantly less with bevy, as the editor objects aren't recreated every frame, but when I drag tiles across the screen, there is a pretty large input lag, that isn't pleasant to look at and gives the user the impression that the app is slow and unresponsive. In the video below i removed the gui part to make sure it doesn't cause the problem. I have found a pull request (#6503) that is merged into main, but switching to that branch doesn't fix the problem. Is there any way this can be reduced or fixed completely or did I miss something about the main branch?

https://user-images.githubusercontent.com/68352124/218064476-1f292f7d-c7b2-4cbe-a720-eea4807bd905.mp4

**Comment by nicopap** [comment, marked as answer]
There is several things you can do to reduce lag:

- Use [bevy framepace](https://lib.rs/crates/bevy_framepace) It's basically a free latency reduction
- Order your systems carefully so that the system responsible for moving the piece runs after the system to select the piece etc. This is very important.

### Setting component values

Another source of latency is inserting/removing components. If you are updating a component's value, you should be using a `Query<&mut MyComponent>` instead of `commands.insert(MyComponent)`.

`Commands` are applied at the end of a stage, so if you have a system that sets `MyComponent`, and another that read its value in the same stage, the reading system will only be able to read the value of `MyComponent` from last frame.

### System ordering

For example, your systems may look like

```rust
app
  .add_system(set_velocity_from_mouse_delta)
  .add_system(update_transform);
```

This runs, but if `set_velocity_from_mouse_delta` runs after `update_transform`. At any given frame, the input of the last frame will be used for the transform.

You need to make sure `set_velocity_from_mouse_delta` runs before `update_transform`, this way:

```rust
app
  .add_system(set_velocity_from_mouse_delta.before(update_transform))
  .add_system(update_transform);
```

**Comment by DKolter** [comment]
Ohhh, thank you very much, i see that there is a lot of room for improvement in my code now. The system ordering is a point i have completely missed, and i am using the commands.insert function every frame for updating the mouse last position. I will look into framepace as well, thank you very much for all these tips and the quick help 💯
